cmd/promtool/analyze.go

```diff
@@ -0,0 +1,543 @@
+	result := []string{}
+	for _, v := range values {
+		s := string(v)
+		if !strings.HasSuffix(s, "_bucket") && !strings.HasSuffix(s, "_count") && !strings.HasSuffix(s, "_sum") {
+			result = append(result, s)
+		}
+	}
+	return result, err
+}
+
+func queryBaseMetricNames(ctx context.Context, api v1.API, start, end time.Time) ([]string, error) {
+	values, _, err := api.LabelValues(ctx, labels.MetricName, []string{"{__name__=~\".*_bucket\"}"}, start, end)
+	if err != nil {
+		return nil, err
+	}
+	result := []string{}
+	for _, v := range values {
+		basename, err := metricBasename(string(v))
+		if err != nil {
+			// Just skip if something doesn't look like a good name.
+			continue
+		}
+		result = append(result, basename)
+	}
+	return result, nil
+}
+
+func metricBasename(bucketMetricName string) (string, error) {
+	if len(bucketMetricName) <= len("_bucket") {
+		return "", fmt.Errorf("potential metric name too short: %s", bucketMetricName)
+	}
+	return bucketMetricName[:len(bucketMetricName)-len("_bucket")], nil
+}
+
+// func queryMetadataForHistograms(ctx context.Context, api v1.API, query, limit string) ([]string, error) {
+// 	result, err := api.Metadata(ctx, query, limit)
+// 	if err != nil {
+// 		return nil, err
+// 	}
+
+// 	metrics := make([]string, 0)
+// 	for metric, metadata := range result {
+// 		for _, metadatum := range metadata {
+// 			if metadatum.Type == "histogram" || metadatum.Type == "gaugehistogram" {
+// 				metrics = append(metrics, metric)
+// 			}
+// 		}
+// 	}
+
+// 	return metrics, nil
+// }
+
+// Query the related count_over_time(*_sum[duration]) series to double check that metricName is a
+// histogram. This keeps the result small (avoids buckets) and the count gives scrape interval
+// when dividing duration with it.
+func queryLabelSets(ctx context.Context, api v1.API, metricName string, end time.Time, duration time.Duration) (model.Vector, error) {
+	query := fmt.Sprintf("count_over_time(%s{}[%s])", metricName, duration.String())
+
+	values, _, err := api.Query(ctx, query, end)
+	if err != nil {
+		return nil, err
+	}
+
+	vector, ok := values.(model.Vector)
+	if !ok {
+		return nil, fmt.Errorf("query for metrics resulted in non-Vector")
+	}
+	return vector, nil
+}
+
+func seriesSelector(bucketMetricName string, lbs model.LabelSet) string {
+	builder := strings.Builder{}
+	builder.WriteString(bucketMetricName)
+	builder.WriteRune('{')
+	first := true
+	for l, v := range lbs {
+		if first {
+			first = false
+		} else {
+			builder.WriteRune(',')
+		}
+		builder.WriteString(string(l))
+		builder.WriteString("=\"")
+		builder.WriteString(string(v))
+		builder.WriteRune('"')
+	}
+	builder.WriteRune('}')
+
+	return builder.String()
+}
+
+func querySamples(ctx context.Context, api v1.API, query string, start, end time.Time, step time.Duration) (model.Matrix, error) {
+	values, _, err := api.QueryRange(ctx, query, v1.Range{Start: start, End: end, Step: step})
```

Comment:
Agreed in histogram weekly sync meeting to use instant query with duration instead, since that automatically returns the actual samples and doesn't need step/scrap interval.